src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala

```diff
@@ -181,35 +207,89 @@ private[impl] class LaunchQueueActor(
         case None => sender() ! None
       }
 
-    case Add(app, count) =>
-      launchers.get(app.id) match {
-        case None =>
-          import context.dispatcher
-          val actorRef = createAppTaskLauncher(app, count)
-          val eventualCount: Future[QueuedInstanceInfo] =
-            (actorRef ? TaskLauncherActor.GetCount).mapTo[QueuedInstanceInfo]
-          eventualCount.map(_ => Done).pipeTo(sender())
+    case add @ Add(runSpec, _) =>
+      // we cannot process more Add requests for one runSpec in parallel because it leads to race condition
+      // the queue handling is helping us ensure we do only one per run-spec at a time
+      // requests for multiple runSpecs are still processed in parallel
 
-        case Some(actorRef) =>
-          import context.dispatcher
-          val eventualCount: Future[QueuedInstanceInfo] =
-            (actorRef ? TaskLauncherActor.AddInstances(app, count)).mapTo[QueuedInstanceInfo]
-          eventualCount.map(_ => Done).pipeTo(sender())
+      val oldQueue: Queue[QueuedAdd] = updatesByRunSpecId(runSpec.id)
+      val newQueue = oldQueue :+ QueuedAdd(sender(), add)
+      updatesByRunSpecId += runSpec.id -> newQueue
+
+      if (oldQueue.isEmpty) {
+        // start processing the just received operation
+        processNextAddIfExists(runSpec)
       }
 
+    case AddFinished(queuedAdd) =>
+      val add = queuedAdd.add
+      val (dequeued, newQueue) = updatesByRunSpecId(add.spec.id).dequeue
+      require(dequeued == queuedAdd)
+      if (newQueue.isEmpty)
+        updatesByRunSpecId -= add.spec.id
+      else
+        updatesByRunSpecId += add.spec.id -> newQueue
+
+      sender() ! Done
+
+      processNextAddIfExists(add.spec)
+
     case msg @ RateLimiterActor.DelayUpdate(app, _) =>
       launchers.get(app.id).foreach(_.forward(msg))
   }
 
-  private[this] def createAppTaskLauncher(app: RunSpec, initialCount: Int): ActorRef = {
-    val actorRef = context.actorOf(runSpecActorProps(app, initialCount), s"$childSerial-${app.id.safePath}")
+  @SuppressWarnings(Array("all")) /* async/await */
+  private def processNextAddIfExists(runSpec: RunSpec): Unit = {
+    import context.dispatcher
+
+    updatesByRunSpecId(runSpec.id).headOption.map { queuedItem =>
+      val future = async {
+        // Trigger TaskLaunchActor creation and sync with instance tracker.
+        val actorRef = launchers.getOrElse(runSpec.id, createAppTaskLauncher(runSpec))
+        val info = await((actorRef ? TaskLauncherActor.Sync(runSpec)).mapTo[QueuedInstanceInfo])
```

Comment:
we don't need this info, just return Done, we don't need the stats

Reply:
add comment that we have to await because TaskLauncherActor reads instancetracker state both directly and via published state events and that state affects the outcome of the sync call